snapcraft/internal/sources/_git.py

```diff
@@ -49,6 +49,11 @@ def _pull_existing(self):
         elif self.source_commit:
             refspec = self.source_commit
 
+        subprocess.check_call([self.command, '-C', self.source_dir,
+                               'fetch', '--recurse-submodules=yes'])
+        subprocess.check_call([self.command, '-C', self.source_dir,
+                               'reset', '--hard', refspec])
+
         # Pull changes to this repository and any submodules.
         subprocess.check_call([self.command, '-C', self.source_dir,
                                'pull', '--recurse-submodules=yes',
```

Comment:
Does the reset recurse over the submodules?  I think I'll need to do something along the lines of what is suggested here: http://stackoverflow.com/a/17871677